### PR TITLE
fix(jit): raise validation error for required fields

### PR DESCRIPTION
**Pull request by karatakis**
**Summary:**  
Fix validation of required fields to cover edge cases for the JIT module

Remember to test it with `cargo test --features force_jit`

**Build & Testing:**

- [X] I ran `cargo test` successfully.
- [X] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [X] I have added relevant unit & integration tests.
- [X] I have updated the [documentation] accordingly.
- [X] I have performed a self-review of my code.
- [X] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

**Comment by karatakis** [comment]
Test case `test-required-fields.md_5.snap` needs discussion. I believe it should not fail, but the current implementation fails. This issue is open for discussion.

**Comment by meskill** [comment]
> Test case `test-required-fields.md_5.snap` needs discussion. I believe it should not fail, but the current implementation fails. This issue is open for discussion.

Why do you think it should not fail? You have non-null list of non-null values and one of the values is null - according to spec the parent value should be resolved to null in that case and that is against the schema

**Comment by karatakis** [comment]
My bad, @meskill; I messed up the numbering, I believe. Maybe I fixed the issue while I was working on it yesterday.

File: tests/execution/test-required-fields.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      id: 1
+      bar: bar_1
+
+# this does not fail
+- request:
+    method: GET
+    url: http://jsonplaceholder.typicode.com/relaxed-missing
+  response:
+    status: 200
+    body: null
```
Comment (meskill) [inline tests/execution/test-required-fields.md:66]: let's add test for `Foo` but with body `id: 1 bar: null`

File: tests/execution/test-required-fields.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+      - id: 2
+        bar: bar_2
+
+# this fails
+- request:
+    method: GET
+    url: http://jsonplaceholder.typicode.com/full-missing
+  response:
+    status: 200
+    body:
```
Comment (meskill) [inline tests/execution/test-required-fields.md:86]: additional test for `body: null`

File: tests/execution/test-required-fields.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+- request:
+    method: GET
+    url: http://jsonplaceholder.typicode.com/inner-missing
+  response:
+    status: 200
+    body:
+      - id: 1
+        bar: bar_1
+      - id: 2
+        bar: null
```
Comment (meskill) [inline tests/execution/test-required-fields.md:109]: for all of the list cases we need probably 4 tests:
- the whole list is null
- the list is not null and some entry is null
- the list is not null and some entry has inner null
- the list with values

wdyt?

File: tests/core/snapshots/test-required-fields.md_7.snap [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+source: tests/core/spec.rs
+expression: response
+---
+{
+  "status": 200,
+  "headers": {
+    "content-type": "application/json"
+  },
+  "body": {
+    "data": null,
```
Comment (meskill) [inline tests/core/snapshots/test-required-fields.md_7.snap:11]: btw if I understand the spec right it should be `data: { innerMissing: null }` here. But I'm not sure is the problem in async_graphql or tailcall. 

Should we leave some todos related to this?

File: src/core/jit/synth/synth.rs [hunk 5]
```diff
@@ -128,10 +142,22 @@ where
...
+                    } else {
+                        Err(ValidationError::ValueRequired.into())
+                    }
+                }
+                crate::core::blueprint::Type::ListType { of_type, non_null: _ } => {
+                    if of_type.is_nullable() {
+                        Ok(Value::null())
+                    } else {
+                        Err(ValidationError::ValueRequired.into())
+                    }
```
Comment (meskill) [inline src/core/jit/synth/synth.rs:159]: this check is everywhere, can we move it somewhere and reuse it?

File: src/core/jit/synth/synth.rs [hunk 6]
```diff
@@ -128,10 +142,22 @@ where
         }
 
         let eval_result = if value.is_null() {
-            if node.type_of.is_nullable() {
-                Ok(Value::null())
-            } else {
-                Err(ValidationError::ValueRequired.into())
+            // link to GraphQL reference https://spec.graphql.org/October2021/#sec-Handling-Field-Errors
+            match &node.type_of {
+                crate::core::blueprint::Type::NamedType { name: _, non_null: _ } => {
```
Comment (meskill) [inline src/core/jit/synth/synth.rs:147]: do we need a match here or we can use `is_nullable` on the type_of itself?

**Review by meskill — CHANGES_REQUESTED** [review]

File: src/core/jit/synth/synth.rs [hunk 7]
```diff
@@ -68,9 +68,9 @@ where
         Ok(Value::object(data))
     }
 
-    /// checks if type_of is an array and value is an array
+    /// checks if both are array or both are scalar type
     #[inline(always)]
-    fn is_array(type_of: &crate::core::blueprint::Type, value: &Value) -> bool {
+    fn is_both_array_or_scalar(type_of: &crate::core::blueprint::Type, value: &Value) -> bool {
```
[the remark this answers is not in the document]
Comment (karatakis) [reply src/core/jit/synth/synth.rs:73]: I renamed it because the original name was not right

[the remark this answers is not in the document]
Comment (meskill) [reply src/core/jit/synth/synth.rs:73]: makes sense

File: src/core/jit/synth/synth.rs (hunk 6 above)
Comment (karatakis) [reply src/core/jit/synth/synth.rs:147]: We want to test the inside of the type in case it is a list.
If we have `[Foo!]` in this scenario, I check if Foo is nullable or not, so I need to unwrap

File: tests/execution/test-required-fields.md (hunk 3 above)
Comment (karatakis) [reply tests/execution/test-required-fields.md:109]: I can do it, it won't take much time. So we can be 100% sure

File: tests/core/snapshots/test-required-fields.md_7.snap (hunk 4 above)
Comment (karatakis) [reply tests/core/snapshots/test-required-fields.md_7.snap:11]: I can check it with the graphql reference implementation and I will reply

Comment (karatakis) [reply tests/core/snapshots/test-required-fields.md_7.snap:11]: yeah data is `{ innerMissing: null }`

Comment (karatakis) [reply tests/core/snapshots/test-required-fields.md_7.snap:11]: Added a TODO, it can be solved to separate issue

File: src/core/blueprint/blueprint.rs [hunk 8]
```diff
@@ -77,6 +77,15 @@ impl Type {
...
         }
     }
+
+    // checks if inner type is nullable
+    pub fn is_inner_nullable(&self) -> bool {
+        match self {
+            Type::NamedType { non_null, .. } => !*non_null,
+            Type::ListType { of_type, .. } => of_type.is_nullable(),
+        }
+    }
```
Comment (tusharmath) [inline src/core/blueprint/blueprint.rs:87]: Technically it should return an `Option<bool>` inner isn't available if self is a NamedType. I would rather have this function inlined into its only usage as this seems more like business logic than a helper method on Type.
Action: Inline the function wherever it's used.

**Review by meskill — APPROVED** [review]
